external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala

```diff
@@ -201,4 +201,25 @@ private[sql] object AvroUtils extends Logging {
       }
     }
   }
+
+  /**
+   * Extract a single field from `avroFields` which has the same name has `catalystField`,
+   * performing the matching with proper case sensitivity according to [[SQLConf.CASE_SENSITIVE]].
+   *
+   * @param avroFields The fields within which to search for a match.
+   * @param catalystField The catalyst field for which to search for a match.
+   * @return `Some(match)` if a matching Avro field is found, otherwise `None`.
+   */
+  private[avro] def getAvroFieldForCatalystField(
+    avroFields: TraversableOnce[Schema.Field],
+    catalystField: StructField): Option[Schema.Field] = {
+    val isEqual: (String, String) => Boolean =
+      if (SQLConf.get.caseSensitiveAnalysis) { _.equals(_) } else { _.equalsIgnoreCase(_) }
+    avroFields.filter(f => isEqual(f.name(), catalystField.name)).toSeq match {
+      case Seq(avroField) => Some(avroField)
+      case Seq() => None
+      case s => throw new IncompatibleSchemaException(
+        s"Searching for ${catalystField.name} in Avro schema gave ${s.size} matches")
```

Comment:
Nit: Let's print the Avro field names in the error message.

Reply:
Also, we need a test case for this

Reply:
Good idea on printing Avro field names! I added a test case as well.